src/Parsing/Impl/Ast/PythonVariable.cs

```diff
@@ -29,12 +31,12 @@ public class PythonVariable {
         /// <summary>
         /// The scope the variable was declared in.
         /// </summary>
-        public ScopeStatement Scope { get; }
+        public IScopeNode Scope { get; }
```

Comment:
Unfortunately this is probably going to be a breaking change. Not entirely certain how bad it'd be, since it seems required to make this work.

Reply:
Yeah I don't see any other way to do it while being backwards compatible, unless you guys have any other ideas

Reply:
The problem is that intellicode refactoring and intellicode extension to LS take dependency on the parser. We can't change it and just ask them to recompile. New LS MUST work with existing IC in the wild. I.e. any release of LS must be compatible with existing IC. IC can be rebuilt guilg forward, but there we can't make coordinated releases.

Basically, if you are making any changes to AST you have to
- install intellicode and verify it works in VS Code
- compile PTVS with new parser or at least coordinate with @huguesv wrt of the changes.

Reply:
I think there's a way, actually. Keep all of the old ones as they were (type + name). Add new ones, and then make the old ones use the new ones. The old ones are concretely typed to be ScopeStatement, so you can traverse the new parent properties up until you hit a scope that is ScopeStatement, and return that.

For example, if something's parent scope is a list comprehension, in the old code it would not have been, but would have been the containing scope of the list, so now it'd walk upward and find the same thing.

Backwards compatible in the API as well as the returned value, I think.

Reply:
```csharp

public ScopeStatement Scope;
public IScopeNode ScopeNode;
```

Here, and then in ScopeStatement:

```csharp
public ScopeStatement Parent;
public IScopeNode ParentNode;
```

Or similar.

Reply:
The problem is that intellicode refactoring and intellicode extension to LS take dependency on the parser. We can't change it and just ask them to recompile. New LS MUST work with existing IC in the wild. I.e. any release of LS must be compatible with existing IC. IC can be rebuilt guilg forward, but there we can't make coordinated releases.

Basically, if you are making any changes to AST you have to
- install intellicode and verify it works in VS Code
- compile PTVS with new parser or at least coordinate with @huguesv wrt of the changes.

Reply:
I think there's a way to do it backwards compatibly.

Put the code back the way it was, and add a new property like `public IScopeNode ScopeNode`. Then make the old one use the new one by walking up the scopes until a scope node of type `ScopeStatement` is found. This means that any extra scopes your code introduces won't be exposed via the old API, so for old users, they'd get the same thing.

Reply:
Then similarly for ScopeStatement, keep the old Parent, then a new ParentNode or something which does the same.

An idea to think about.

Reply:
Ugh, github glitched and I thought it didn't send my comments, so I typed out two replies... Oops

Reply:
Ok I'll give it a shot

Reply:
I think we will still have to do the steps mikhail talked about, but I tried your suggested changes. It doesn't look quite as good but we can always revert it.

Reply:
The testing? Definitely. Testing with IntelliCode first for sure.

Reply:
Could you walk me through on how to do that, I'm not really sure what you guys mean by what version of intellicode and how to compile PTVS.

Reply:
For IntelliCode, basically just install this and run the LS and see if completion breaks. https://marketplace.visualstudio.com/items?itemName=VisualStudioExptTeam.vscodeintellicode

Reply:
Ok I'll give it a shot

Reply:
So I tried it out and I just see 
```
Acquiring model 'intellisense-members' for python
Querying IntelliCode service for available models.
Cached model is up to date.
vs-intellicode-python was passed a model: {"modelPath":"/home/ctrando/.vscode/extensions/visualstudioexptteam.vscodeintellicode-1.1.9/cache/88630A1B4C077752096ECAA576B92C7D3FA1_BC98A65409D042A99DDAF47DC30E1530","modelId":"88630A1B4C077752096ECAA576B92C7D3FA1","outputId":"BC98A65409D042A99DDAF47DC30E1530"}.
Loaded language server extension.
```

Reply:
You'd need to go and try out some completions that fit their model. You'll know they're being sorted if you see a star next to the items in the completion list.

Reply:
Ok I verified that I still get star stuff with my changes

Reply:
That's excellent. I think you've worked this to be API backwards compatible.

Reply:
I can barely take any credit since you helped so much with it! But you should try and verify on your end as well. Let me know if there's anything else you want me to do.